ui/app/components/clients/line-chart.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import Component from '@glimmer/component';
+import { SVG_DIMENSIONS, formatNumbers } from 'vault/utils/chart-helpers';
+import { parseAPITimestamp } from 'core/utils/date-formatters';
+import { format, isValid, parse } from 'date-fns';
+import { SerializedChartData, UpgradeData } from 'vault/client-counts';
+import { debug } from '@ember/debug';
+
+interface Args {
+  dataset: SerializedChartData[];
+  upgradeData: UpgradeData[];
+  xKey?: string;
+  yKey?: string;
+  chartHeight?: number;
+  dateFormat?: string;
+}
+
+interface ChartData {
+  x: Date;
+  y: number | null;
+  new: number;
+  tooltipUpgrade: string | null;
+}
+
+interface UpgradeByMonth {
+  [key: string]: UpgradeData;
+}
+
+/**
+ * @module LineChart
+ * LineChart components are used to display time-based data in a line plot with accompanying tooltip
+ *
+ * @example
+ * ```js
+ * <LineChart @dataset={{dataset}} @upgradeData={{this.versionHistory}}/>
+ * ```
+ * @param {string} xKey - string denoting key for x-axis data of dataset. Should reference a date string with format 'M/yy'.
+ * @param {string} yKey - string denoting key for y-axis data of dataset. Should reference a number or null.
+ * @param {array} upgradeData - array of objects containing version history from the /version-history endpoint
+ * @param {number} [chartHeight=190] - height of chart in pixels
+ * @param {string} [dateFormat='M/yy'] - format of date string in xKey
+ */
+export default class LineChart extends Component<Args> {
+  // Chart settings
+  get yKey() {
+    return this.args.yKey || 'clients';
+  }
+  get xKey() {
+    return this.args.xKey || 'month';
+  }
+  get chartHeight() {
+    return this.args.chartHeight || SVG_DIMENSIONS.height;
+  }
+  get dateFormat() {
+    return this.args.dateFormat || 'M/yy';
+  }
+  // Plot points
+  get data(): ChartData[] {
+    try {
+      return this.args.dataset?.map((datum) => {
+        const timestamp = parse(datum[this.xKey] as string, this.dateFormat, new Date()) as Date;
+        if (isValid(timestamp) === false)
+          throw new Error(
+            `Unable to parse value "${datum[this.xKey]}" as date with format ${this.dateFormat}`
+          );
+        const upgradeMessage = this.getUpgradeMessage(datum);
+        return {
+          month: datum[this.xKey],
+          x: timestamp,
+          y: (datum[this.yKey] as number) ?? null,
+          new: this.getNewClients(datum),
+          tooltipUpgrade: upgradeMessage,
+        };
+      });
+    } catch (e) {
+      debug(e as string);
+      return [];
+    }
+  }
+  get upgradedMonths() {
+    return this.data.filter((datum) => datum.tooltipUpgrade);
+  }
+  // Domains
+  get yDomain() {
+    const setMax = Math.max(...this.data.map((datum) => datum.y ?? 0));
+    const nearest = setMax < 1500 ? 200 : 2000;
+    // Round to upper 200 or 2000
+    return [0, Math.ceil(setMax / nearest) * nearest];
+  }
```

Comment:
Does this mean that the chart will have a hardcoded max y value of 200? This seems a bit strange if it's a new vault instance and a customer has fewer than 100 clients. 

I'd think we'd want to use `ScaleLinear` here? 

And below we could use `ScaleTime`

Reply:
No, it rounds to the nearest 200 or 2000 (for larger values). This is basically to make sure we don't have a scale of like 0 - 1 for smaller counts